### PR TITLE
U4-7360 - Toggle password in login screen

### PR DESCRIPTION
**Pull request by bjarnef**
Add an icon to toggle the password visibility (show/hide) in login screen.
http://issues.umbraco.org/issue/U4-7360

Umbraco 7 includes the font icons from http://hlvticons.ch/ .. however it only include an "eye" icon, not and "eye-off", like in the font set here http://zavoloklom.github.io/material-design-iconic-font/icons.html that I have used for the package https://our.umbraco.org/projects/backoffice-extensions/material-design-icon-pack/) or "eye-close" included in font awesome http://fortawesome.github.io/Font-Awesome/3.2.1/icon/eye-close/

Maybe the creators from http://hlvticons.ch/ can add a similar one to their font set, so it can be included in the Umbraco 7 core font set?

image

So that is why the icon is missing for now when I click "show password".
image

### CODE REVIEW

File: src/Umbraco.Web.UI.Client/src/views/common/dialogs/login.html [hunk 1]
```diff
@@ -40,7 +40,10 @@
                 </div>
 
                 <div class="control-group" ng-class="{error: loginForm.password.$invalid}">
-                    <input type="password" ng-model="password" name="password" class="input-xlarge" localize="placeholder" placeholder="@placeholders_password" autocomplete="off" />
+                    <input type="{{showpassword ? 'text' : 'password'}}" ng-model="password" name="password" class="input-xlarge password" ng-class="{'has-password': password.length > 0}" localize="placeholder" placeholder="@placeholders_password" autocomplete="off" />
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/views/common/dialogs/login.html]: We need to double check that setting 'type' dynamically with interpolation will work in all browsers. For example, interpolated attributes in IE for things like 'class' don't work. You might have to use something like ng-attr-type to be sure.

Comment (bjarnef) [inline src/Umbraco.Web.UI.Client/src/views/common/dialogs/login.html]: We could check with this example http://codepen.io/gymbry/full/fJchw/ which use the same principle.. however I just noticed that the Microsoft Edge browser actually add this feature to password input fields.

image

the same for IE11 (at least on Windows 10)
image

Comment (bjarnef) [inline src/Umbraco.Web.UI.Client/src/views/common/dialogs/login.html]: Okay, this seems to work in IE the hide these buttons:

```
input::-ms-clear, input::-ms-reveal{
    display:none;
}
```

Comment (bjarnef) [inline src/Umbraco.Web.UI.Client/src/views/common/dialogs/login.html]: @Shazwazza I just saw this example using a directive:
http://angularscript.com/angular-toggle-password-directive/

demo: http://codepen.io/jakob-e/pen/eNBQaP

**Comment by warrenbuckley** [comment]
Hello @bjarnef 
Sorry for the delayed response on this PR. We are trying our best to be get round to PRs.

Can I get the latest status on this PR. Do you need help from us to finish it off or are you still investigating swapping it out with the directive you found?

If this is ready to be reviewed as a PR to consider then please let us know.

Thanks,
Warren :)

**Comment by bjarnef** [comment]
Hi @warrenbuckley 

There are a few things we need to change.
E.g. the font icons in core only inlcude an eye icon and not an eye-off icon like in font awesome:
http://fontawesome.io/icon/eye/
http://fontawesome.io/icon/eye-slash/

Some browsers do has a built-in feature for show/hide password, e.g. IE and Edge (like some browsers also display a clear icon for `<input type="search">` and spinner controlls for `<input** type="number">`) - should we just hide the default UI controls for these browsers?

```
input::-ms-clear, input::-ms-reveal{
    display:none;
}
```

Does it has any issues, when switching between password/text like @Shazwazza mentioned?

**Comment by warrenbuckley** [comment]
OK we would need to see what we can do in regards to getting another icon-sorted. Is there nothing else in the iconset that we ship with that could be used as an alternative @bjarnef ?

If we are showing custom UI to do the same thing as some browsers built in functionality, myself personally say that we should indeed hide the default browser controls to avoid confusion between them.

In regards to @Shazwazza we will need to just perform lots of browser tests to ensure that switching the input type attribute with Angular works in all browsers we support.

What have you tested this PR on with so far?

**Comment by bjarnef** [comment]
@warrenbuckley well I think that is these icons most people are using for show/hide password feature. I guess people are familier with these.
E.g. https://login.mailchimp.com/

Is it possible that Umbraco can include a larger set of icons from http://hlvticons.ch/ ?

I tested the PR in some of the most popular browsers like Chrome, Firefox, IE 11 and Edge.

**Comment by bjarnef** [comment]
Some are using text instead of the icon or combine text with a checkbox.
Not sure how well it will work with text, if it should be located "inside" the textbox.
https://codepen.io/geoffreyrose/pen/HKDkB

Some alternatives:
https://codepen.io/jasesmith/pen/WRXoRP
https://ux.stackexchange.com/questions/54162/icon-for-unmasking-passwords-open-or-closed-to-begin-with

**Comment by bjarnef** [comment]
The demo here http://simpulton.github.io/angularjs-mailchimp-signup/ use same approach to switch between password/text and seems to work in IE9+

http://onehungrymind.com/build-mailchimp-signup-form-angularjs/

**Comment by zpqrtbnk** [comment]
Have re-implemented using an Angular directive inspired from the one you posted at some point.
In PR https://github.com/umbraco/Umbraco-CMS/pull/2246

Works quite well for me, suggesting we close this PR, moving the other one for review.

**Comment by zpqrtbnk** [comment]
Other PR has been merged, and issue fixed, closing this one.

**Pull request closed without merging**